### DISCUSSION TITLE
Make Filter decorator behave more like scope in rails

### DISCUSSION THREAD

**Discussion by YegorMy** [mikro-orm/mikro-orm#4654] [Ideas]
**Is your feature request related to a problem? Please describe.**
I was working with RoR for quite some time and a concept of scopes now lives in my head. So it would be nice if we had somewhat resembling that in mikro-orm now.

**Describe the solution you'd like**

### Schema Definition

```typescript
@Filter({ name: 'active', cond: { inactiveAt: null } })
@Entity()
export class UserEntity  {
  @Property() public firstName!: string;
  @Property() public lastName!: string;
  @Propery({ nullable: true }) public inactiveAt: string | null;

  @OneToMany(() => UserGroup, e => e.user) public groups = new Collection<UserGroup>(this);
}

@Filter({ name: 'active', cond: { inactiveAt: null } })
@Entity()
export class UserGroupEntity  {
  @Property() public name!: string;
  @Property() public displayName!: string;
  @Propery({ nullable: true }) public inactiveAt: string | null;

  @ManyToOne(() => User) public user!: UserEntity;
}
```

### Usage
```typescript
const user = await entityManager.find(UserEntity, {}, { populate: { groups: 'active' } }); // Will return only active groups for the user
```

So that would apply a scope filter to the whole subquery making it somewhat like this:

```mysql
select u.*, g.* from user_entity u left join user_group_entity g on g.user_id = u.id where g.inactive_at is null;
```

**Describe alternatives you've considered**
So the alternative to this solution is to write a repository for an entity and create functions to fetch that yourself, like:

```typescript
entityManager
      .qb(UserEntity, 'u')
      .select('*')
      .leftJoinAndSelect('u.groups', 'g', { 'g.inactiveAt': null });
```

**Additional context**
So this whole idea is way greater than just having a 3NF fields and `active` filters. If you were able to stack those deeper you go would be absolutely awesome.
But I suppose that would mean `@Filter` decorator should be eradicated for the favour of something like

```typescript
@Entity()
export class UserEntity extends Entity {
  @Scope('active') protected get activeScope() { return { inactiveAt: null } }
}
```

so we would be able to properly autocomplete it.

I suggest nesting should also be able to take additional parameters like, so we can add up to the scope or filter as we go.
Like
```
await entityManager.find(UserEntity, {}, { populate: { groups: { filters: ['active'], where: { name: { like: '%awesome orm%' } } } });
```

Thanks for the amount of work you did, I'm really grateful for this ORM!

**Comment by B4nan** [comment]
Never used rails myself, will need to check how the scopes work, maybe they could live with filters side by side as an alternative approach. 

Speaking of the proposed API, I'm afraid the populate parameter needs to stay as string array, mainly to keep the inference between it and the `Loaded` return type working, but we could find some other way for sure.

**Comment by YegorMy** [comment]
It's not really that important on how it is going to be implemented. The concept itself really makes the life way easier and most importantly renders simple find functions in repositories or finders obsolete.

Link to the guide.
https://www.rubyguides.com/2019/10/scopes-in-ruby-on-rails/

**Comment by B4nan** [comment]
Have you seen #2440 and #2434? This seems a bit like a duplicate of those, as the only missing thing should be filters on relations, otherwise you can in fact operate with them the same way as with scopes I think. Hard to compare this with other languages, I doubt ruby has this type-safe (idk, never used ruby), while things are possible on runtime, the tricky part is to come up with a sane API that can be type-safe.

I know the problem is there, but so far nobody suggested a good way to go about it (API-wise) that would be type-safe and won't bring more problems (e.g. your proposal would break the strict relations completely).

**Comment by 1gluk** [comment]
try this variant:
`@Filter({ name: 'active', cond: { inactiveAt: null,  groups: { inactiveAt: null}} })`
and
`await entityManager.find(UserEntity, {}, {
            filters:  {...}, /* active and etc */
            populate: ['groups'], 
            populateWhere: { 
                        groups: {
                                     name: {
                                                 $like: '%awesome orm%'
                                                 }
                                     }
                                     }
                         }
            );`
